packages/opentelemetry-instrumentation-http/src/http.ts

```diff
@@ -535,8 +524,19 @@ export class HttpInstrumentation extends InstrumentationBase<Http> {
       }
 
       const operationName = `${component.toUpperCase()} ${method}`;
+
+      const hostname =
+        optionsParsed.hostname ||
+        optionsParsed.host?.replace(/^(.*)(:[0-9]{1,5})/, '$1') ||
```

Comment:
I think this could be problematic if someone  uses an ipv6 address like `::1`.

Reply:
This is actually not a new code, I just moved it from the response callback to execute before the span start.

You are right that the following invocation will not work correctly:
```
> '::1'.replace(/^(.*)(:[0-9]{1,5})/, '$1')
':'
```

I am not very familiar with ipv6, but when I tried to execute`http.get` like this (with no instrumentation), it did not work:
```js
http.get('http://::1', (res) => {
    res.on('data', () => console.log('got data'));
    res.on('end', () => console.log('res end'));
});

[ERROR] 11:47:05 TypeError: Invalid URL: http://::1
```

While the following did work:
```js
http.get('http://[::1]', (res) => {
    res.on('data', () => console.log('got data'));
    res.on('end', () => console.log('res end'));
});

// prints: 
got data
res end
```

and for the working example, the span is generated correctly (as the IP is surrounded with []):
```js
{
  traceId: '0e79148c5f33c65def42317ca86ff430',
  parentId: undefined,
  name: 'HTTP GET',
  id: '37ea719c7f94f4ac',
  kind: 2,
  timestamp: 1626338943075430,
  duration: 22224,
  attributes: {
    'http.url': 'http://[::1]/',
    'http.method': 'GET',
    'http.target': '/',
    'net.peer.name': '::1',
    'net.peer.ip': '::1',
    'net.peer.port': 80,
    'http.host': '::1:80',
    'http.response_content_length_uncompressed': 19,
    'http.status_code': 404,
    'http.status_text': 'NOT FOUND',
    'http.flavor': '1.1',
    'net.transport': 'ip_tcp'
  },
  status: { code: 2 },
  events: []
}
```

It also works for invocation with ipv6 and a port (although the `http.host` is lacking the brackets which is not consistent with the `http.url`, not sure if OK or a problem):
```js
http.get('http://[::1]:8080', (res) => {
    res.on('data', () => console.log('got data'));
    res.on('end', () => console.log('res end'));
});

{
  traceId: 'a5a8b4b6c42a781bf924b13181fb0f79',
  parentId: undefined,
  name: 'HTTP GET',
  id: '92a6b69a6c184f2a',
  kind: 2,
  timestamp: 1626339282167529,
  duration: 18508,
  attributes: {
    'http.url': 'http://[::1]:8080/',
    'http.method': 'GET',
    'http.target': '/',
    'net.peer.name': '::1',
    'net.peer.ip': '::1',
    'net.peer.port': 8080,
    'http.host': '::1:8080',
    'http.response_content_length_uncompressed': 2,
    'http.status_code': 200,
    'http.status_text': 'OK',
    'http.flavor': '1.1',
    'net.transport': 'ip_tcp'
  },
  status: { code: 1 },
  events: []
}
```

As the ipv6 issue is not in the scope of this PR, and I tried to find a failing example and was not able to, I think it's best if we create a separate issue for it if a problematic example is found.

WDYT?

Reply:
You are right, that's existing code. Haven't noticed this during reviewing.
I haven't tried but I was thinking about following:
```js
http.get({ host: "::1" })
```

Reply:
you are right!
It does not work

```js
http.get({host: '::1', port: 8080}, (res) => {
    res.on('data', () => console.log('got data'));
    res.on('end', () => console.log('res end'));
});
```

produces:
```js
{
  traceId: 'cdd140118cdfc23da00059c291a18b7a',
  parentId: undefined,
  name: 'HTTP GET',
  id: '9aeb488a5ff193b6',
  kind: 2,
  timestamp: 1626344909735352,
  duration: 29152,
  attributes: {
    'http.url': 'http://::1/',
    'http.method': 'GET',
    'http.target': '/',
    'net.peer.name': ':',
    'net.peer.ip': '::1',
    'net.peer.port': 8080,
    'http.host': '::8080',
    'http.response_content_length_uncompressed': 2,
    'http.status_code': 200,
    'http.status_text': 'OK',
    'http.flavor': '1.1',
    'net.transport': 'ip_tcp'
  },
  status: { code: 1 },
  events: []
}
```

and "http.host", and "http.url" are indeed wrong.
I'll open an issue for that, good catch :)

Reply:
opened an issue for this bug:
https://github.com/open-telemetry/opentelemetry-js/issues/2353

Is it ok if I resolve the conversation for this PR? as this is an existing bug, not related to my changes, and out of the scope for this PR